aptos-move/dynamic-transaction-composer/src/codegen.rs

```diff
@@ -0,0 +1,380 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    ArgumentOperation, BatchArgument, BatchedFunctionCall, PreviousResult, APTOS_SCRIPT_BUILDER_KEY,
+};
+use move_binary_format::{
+    access::{ModuleAccess, ScriptAccess},
+    builders::CompiledScriptBuilder,
+    errors::{PartialVMError, PartialVMResult},
+    file_format::*,
+};
+use move_core_types::{
+    identifier::IdentStr,
+    language_storage::{ModuleId, TypeTag},
+    metadata::Metadata,
+    transaction_argument::TransactionArgument,
+    vm_status::StatusCode,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+#[derive(Default)]
+struct Context {
+    script: CompiledScript,
+    // Arguments to the compiled script.
+    args: Vec<Vec<u8>>,
+    // (nth move call) -> (type arguments for the call)
+    ty_args: Vec<Vec<SignatureToken>>,
+    // (nth move call) -> (starting position of its return values in the local pool)
+    returned_val_to_local: Vec<u16>,
+    // (nth move call) -> (starting position of its arguments in the argument pool)
+    args_to_local: Vec<u16>,
+    // (nth move call) -> how many return values it got
+    return_counts: Vec<u16>,
+    // Number of signers at the beginning of arugment pool
+    signer_counts: u16,
+    // Local signature pool of the script. We will use this pool to replace the pool in CompiledScript.
+    locals: Vec<SignatureToken>,
+    // Parameter signature pool of the script. We will use this pool to replace the pool in CompiledScript.
+    parameters: Vec<SignatureToken>,
+}
+
+fn instantiate(
```

Comment:
Why not right away move this into `SignatureToken`. Seems to be generic and useful enough.